src/iosrv/srv.c

```diff
@@ -233,6 +233,122 @@ dss_sched_unit_pop(ABT_pool *pools, ABT_pool *pool)
 	return ABT_UNIT_NULL;
 }
 
+static struct dss_xstream *
+dss_xstream_get(int stream_id)
+{
+	if (stream_id == DSS_XS_SELF)
+		return dss_get_module_info()->dmi_xstream;
+
+	D_ASSERTF(stream_id >= 0 && stream_id < xstream_data.xd_xs_nr,
+		  "invalid stream id %d (xstream_data.xd_xs_nr %d).\n",
+		  stream_id, xstream_data.xd_xs_nr);
+
+	return xstream_data.xd_xs_ptrs[stream_id];
+}
+
+/* Add to the sorted(by expire time) list */
+static void
+dss_add_expire_list(struct dss_xstream *dx, struct dss_timeout_ult *new)
+{
+	struct dss_timeout_ult	*dtu;
+
+	d_list_for_each_entry(dtu, &dx->dx_timeout_ult_list, dtu_list) {
+		if (dtu->dtu_expire_time > new->dtu_expire_time) {
+			d_list_add_tail(&new->dtu_list, &dtu->dtu_list);
+			return;
+		}
+	}
+
+	d_list_add_tail(&new->dtu_list, &dx->dx_timeout_ult_list);
+}
+
+struct dss_timeout_ult
+*dss_timeout_create(void)
+{
+	struct dss_timeout_ult *dtu;
+	ABT_thread	     self;
+
+	D_ALLOC_PTR(dtu);
+	if (dtu == NULL)
+		return NULL;
+
+	ABT_thread_self(&self);
+	dtu->dtu_expire_time = 0;
+	dtu->dtu_thread = self;
+	D_INIT_LIST_HEAD(&dtu->dtu_list);
+
+	return dtu;
+}
+
+static void
+dss_timeout_internal(struct dss_timeout_ult *dtu, uint64_t timeout)
+{
+	struct dss_xstream	*dx = dss_xstream_get(DSS_XS_SELF);
+	uint64_t		now = 0;
+
+	/* If the xstream is existing, then the caller needs to
+	 * check that and stop retrying.
+	 */
+	if (dss_xstream_exiting(dx))
+		return;
```

Comment:
This looks not necessary to me, and dss_timeout_internal() could be merged to dss_sched_timeout()/dss_ult_sleep().

Reply:
ok